pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		if !inRange(index, m.Entries) {
+			return true
+		}
+		tracedEntries = append(tracedEntries,
+			kvserverpb.TracedEntry{
+				Index:   index,
+				TraceID: t.traceID,
+				SpanID:  t.spanID,
+			},
+		)
+		// TODO(baptist): Not all the fields are relevant to log for all
+		// message types. Consider cleaning up what is logged.
+		t.logf(4,
+			"%s->%s %v Term:%d Log:%d/%d",
+			raft.DescribeTarget(m.From),
+			raft.DescribeTarget(m.To),
+			m.Type,
+			m.Term,
+			m.LogTerm,
+			m.Index,
+		)
+		return true
+	})
+	return tracedEntries
+}
+
+// inRange returns true if the index is within the range of the entries. It
+// assumes the entries are sorted by index. Specifically it returns if the index
+// is in the set [first_index,last_index].
+func inRange(index kvpb.RaftIndex, entries []raftpb.Entry) bool {
+	return len(entries) != 0 &&
+		index >= kvpb.RaftIndex(entries[0].Index) &&
+		index <= kvpb.RaftIndex(entries[len(entries)-1].Index)
+}
+
+// traceIfPast will log the message the message is past any registered tracing
+// points. It will additionally unregister traces that are no longer useful.
+// This call is for events that move the needle/watermark forward (e.g. the log
+// storage syncs), but don't have an exact range of entries affected. So, being
+// unable to match these events to entries exactly once, we instead check that
+// the watermark passed the entry. To protect against overly verbose logging, we
+// only allow MsgAppResp and MsgStorageAppendResp to be logged once per trace.
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgAppResp:
+			if kvpb.RaftIndex(m.Index) >= index && t.appRespSent.CompareAndSwap(false, true) {
+				t.logf(4,
+					"%s->%s %v Term:%d Index:%d",
+					raft.DescribeTarget(m.From),
+					raft.DescribeTarget(m.To),
+					m.Type,
+					m.Term,
+					m.Index,
+				)
+			}
+		case raftpb.MsgStorageAppendResp:
+			if kvpb.RaftIndex(m.Index) >= index && t.storageAppendRespSent.CompareAndSwap(false, true) {
+				t.logf(4,
+					"%s->%s %v Log:%d/%d",
+					raft.DescribeTarget(m.From),
+					raft.DescribeTarget(m.To),
+					m.Type,
+					m.LogTerm,
+					m.Index,
+				)
+			}
+		case raftpb.MsgStorageApplyResp:
+			if len(m.Entries) == 0 {
+				return true
+			}
+			// Use the last entry to determine if we should log this message.
+			msgIndex := m.Entries[len(m.Entries)-1].Index
+			if kvpb.RaftIndex(msgIndex) >= index {
+				t.logf(4,
+					"%s->%s %v Term:%d Index:%d",
+					raft.DescribeTarget(m.From),
+					raft.DescribeTarget(m.To),
+					m.Type,
+					m.Entries[len(m.Entries)-1].Term,
+					msgIndex,
+				)
+				// We unregister the index here because we are now "done" with
+				// this and don't expect more useful events. There could still
+				// be non-quorum messages, but they don't impact anything the
```

Comment:
The non-quorum messages thing does not apply here, suggest removing the last sentence.

This logging happens on all replicas. The "quorum" reasoning is relevant only on the leader.

Reply:
Removed.